### ISSUE TITLE
Smart-case tab completion has become less smart

### ISSUE THREAD

**Issue by zx8** [fish-shell/fish-shell#7944]
Using https://github.com/fish-shell/fish-shell/commit/9c413b039d572963dd5cf9a673ec47927d584df2.

---

It's unclear if this is intentional, but it seems like a regression in smart-case tab completion:

### Repro

```
$ mkdir dog Duck Dodo
$ cd do<tab>
```

### Expected

Both `dog` and `Dodo` should be offered`

### Actual

`dog` is automatically completed - `Dodo` is not offered.

### Rationale

When using lowercase and tab-completing, it's because I can't remember which directories are lowercase and which are uppercase. When I use uppercase (i.e. `cd D<tab>`), it's because I _know_ a particular directory starts with an uppercase letter and going out of my way to enter an uppercase character.

This would be consistent with how smart-case works with all other CLI tools I use, so I expect this is a regression (since it didn't use to work this way in fish either).

**Comment by zanchey** [comment]
I don't know about a regression - this is how 3.0.2 and 3.1.2 behave as well. `complete -C` shows the right result, but something else removes `Dodo` as a valid completion when using <kbd>Tab</kbd>.

**Comment by zanchey** [comment]
See also https://github.com/fish-shell/fish-shell/issues/3978

**Comment by krobelus** [comment]
In the interactive reader [only the best completions are shown](https://github.com/fish-shell/fish-shell/blob/2fff12104d7ea37323beb266a030ff89610d1ee2/src/reader.cpp#L1947-L1948). The rankings are described in https://github.com/fish-shell/fish-shell/issues/568#issuecomment-18455181.
The match type (exact/prefix/substring/subsequence) [dominates the case differences](https://github.com/fish-shell/fish-shell/blob/master/src/wcstringutil.cpp#L212).

**Comment by krackers** [comment]
Note that even if the above is fixed (either by ranking smart-case and case-sensitive the same or by changing the reader to permit ranks that differ by case only), there'd still be an issue where the reader rejects smartcase completions because they require replacement

>If any commands of the best rank do not require replacement, then ignore all those that want to use replacement.


@krobelus Do you know why there exists a distinction between completions that replace the token and ones that don't? From a user perspective why not just treat all of them as replacing token (which would be a visual no-op if the prefix already matches?) Is this just a performance optimization?

**Comment by krobelus** [comment]
> Do you know why there exists a distinction between completions that replace the token and ones that don't?

Right, the distinction only exists internally and does not show up in `complete -C` output.
We should probably try to always use the COMPLETE_REPLACES_TOKEN behavior.

A previous attempt that did something similar was
b38a23a46 (Attempt to simplify how completions get presented in the pager, 2020-11-28)
but reverted in
b9b84e63b (Revert "Attempt to simplify how completions get presented in the pager", 2020-12-02)

To avoid similar issues we should often
- try to avoid modifying what the user typed
- still escape whatever we add

This might be tricky to reconcile with our existing expand.cpp code.

For example, on

	touch file\ with\ spaces
	echo file\ with|

completion should give

	echo file\ with\ spaces

**Comment by krackers** [comment]
@krobelus

Hm what if we modify b38a23a by doing something like

```
   // Ensure that all surviving completions replace their token, so we can handle them uniformly.
    for (completion_t &c : surviving_completions) {
        if (!(c.flags & COMPLETE_REPLACES_TOKEN)) {
            c.flags |= COMPLETE_REPLACES_TOKEN;
            // If current user input includes characters that should not be escaped
            if (escape(tok) != tok) {
                c.flags |= COMPLETE_DONT_ESCAPE;
               c.completion = tok + escape(c.completion,);
            }
            else {
                c.completion.insert(0, tok);
            }
        }
    }
```

This satisfies both the criteria you mentioned.

**Comment by mike-lloyd03** [comment]
I'm migrating to `fish` from `zsh`. This thread is a bit old. Has this issue been addressed in elsewhere?

**Comment by ridiculousfish** [comment]
Doesn't look like it.

**Comment by archelaus** [comment]
I'm surprised there ain't much activity here. I've been noticing this issue since quite a while now, and frankly it's annoying. Like the OP mentioned it's hard to remember off the top of my head whether or not a particular directory starts with a uppercase letter, so IMO tab completion should suggest case-insensitive options as well. Is this something that's being worked on, or has been fixed but I missed it somehow? I <3 Fish but this particular choice seems strange, even Bash doesn't seem to have this issue.

**Comment by TonyWael** [comment]
> I'm surprised there ain't much activity here. I've been noticing this issue since quite a while now, and frankly it's annoying. Like the OP mentioned it's hard to remember off the top of my head whether or not a particular directory starts with a uppercase letter, so IMO tab completion should suggest case-insensitive options as well. Is this something that's being worked on, or has been fixed but I missed it somehow? I <3 Fish but this particular choice seems strange, even Bash doesn't seem to have this issue.

I have to be very honest with you and I totally feel the same. see this behaviour also:

<img width="777" alt="image" src="https://github.com/fish-shell/fish-shell/assets/3015633/1b4c3984-8c30-427e-99b8-48386dea2e0c">

**Comment by limaceous-bushwhacker** [comment]
> The loss of underline is only a minor visual issue, the more major one with actual functionality loss is that it removes the code to automatically insert the common prefix
> 
> The version of my patch based on ridiculousfish's approach actually does preserve both the shared prefix insertion as well as the underline. I tested on your example and it produces the results you expect in your screenshot. Note that `common_prefix` is passed as an arg to `set_completions` which is updated according to the logic in ridiculousfish's patch. His logic to determine this `common_prefix` was also updated to take into account smartcase behavior, see https://github.com/krackers/fish-shell/blob/3.3.1/src/reader.cpp#L1929
> 
> Ridiculousfish's original patch did not properly handle escapes though, which is what my attempt tries to hackily fix. The most robust method I could think of was to still use the approach of uniformly using `complete_replaces_token` (by concating the suggestion to the current token if the replace bit was not set) for populating suggestions, but when inserting suggestions we reconstruct the original suffix if needed from the concatenatd version and only escape that suffix. This feels really hacky though, and there was already one edge-case I had to fix (when you insert a prefix, the current token changes, so you have to explicitly record the beginning of the suffix instead of implicitly getting it by chopping off the current token).
> 
> I briefly looked through your patch razzius, and the line ` c.completion = tok + c.completion;` seems a bit suspect to me because it means that when inserting `tok` might get escaped when it shouldn't be, I think the same issue that ridiculousfish's original patch ran into. I'm not 100% sure though...

Hey, is there a (hopefully, beginner friendly) way to use your patch? Doesn't look there's been much progress in here officially.

**Comment by jdanbrown** [comment]
👋 Bumping so this issue isn't forgotten — it's been 1y since the last update.

As a user migrating from bash to fish I find this behavior surprising and confusing. For example, if you have an iOS project named `Tomato` then you end up with mixed-case files and dirs like this:
```sh
$ ls -1F
Tomato/
Tomato.xcodeproj/
tomato.code-workspace
```

And tab-completing on `tom` only gives the lowercase completion without the uppercase completions, which I find surprising:
```sh
$ ls tom<tab>
$ ls tomato.code-workspace
```
